tests/test_cases/flock.c

```diff
@@ -0,0 +1,77 @@
+#undef _GNU_SOURCE
+#define _GNU_SOURCE
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/fcntl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+// Helper function to remove a file
+int remove_file(char file_name[]) {
+  int rm = remove(file_name);
+  if (rm == -1) {
+    perror("FAILED TO REMOVE FILE\n");
+    exit(EXIT_FAILURE);
+  }
+  return 0;
+}
+
+int main(int argc, char **argv) {
+  // Create a temporary file to use
+  char file_name[100] = "testfiles/flock.txt";
+  int fd = open(file_name, O_RDWR | O_CREAT, 0777);
+  if (fd == -1) {
+    perror("OPEN FAILED\n");
+    exit(EXIT_FAILURE);
+  }
+
+  // Get an exclusive lock on the file
+  int lock = flock(fd, LOCK_EX);
+  if (lock == -1) {
+    perror("FAILED TO GET EXCLUSIVE LOCK\n");
+    close(fd);
+    remove_file(file_name);
+    exit(EXIT_FAILURE);
+  }
+  printf("Obtained exclusive lock on file\n");
+
+  // Release the exclusive lock on the file
+  lock = flock(fd, LOCK_UN);
+  if (lock == -1) {
+    perror("FAILED TO RELEASE EXCLUSIVE LOCK\n");
+    close(fd);
+    remove_file(file_name);
+    exit(EXIT_FAILURE);
+  }
+  printf("Released exclusive lock on file\n");
+
+  // Get a shared lock on the file
+  lock = flock(fd, LOCK_SH);
```

Comment:
@kpan2034 
I am not sure of this but can we modify the test to have 2 shared locks with read permissions on the same file so that we can test that it is possible to create multiple shared locks.

Reply:
I think I should be able to do this by forking the process before acquiring a shared lock. Let me check this out and get back to you.

Reply:
I've updated the test to add the following check:

The parent process first acquires a shared lock, and then spawns another process.
The child process should also be able to successfully acquire a shared lock on the same process.

Also, I just wanted to mention that this question led to me finding a bug in our flock implementation, which I've documented here: https://github.com/Lind-Project/safeposix-rust/issues/185